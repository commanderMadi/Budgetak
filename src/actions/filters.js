// SET_TEXT_FILTER ACTION
export const setTextFilter = (text = '') => ({
    type: 'SET_TEXT_FILTER',
    text
})

//SORT_BY_AMOUNT Action

export const sortByAmount = () => ({
    type: 'SORT_BY_AMOUNT'
})

//SORT_BY_DATE Action

export const sortByDate = () => ({
    type: 'SORT_BY_DATE'
});

//SET_START_DATE Action
export const setStartDate = (date = undefined) => ({
    type: 'SET_START_DATE',
    date
})

export const setEndDate = (date = undefined) => ({
    type: 'SET_END_DATE',
    date
})

