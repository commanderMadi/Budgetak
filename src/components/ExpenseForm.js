import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            err: ''
        }
    }

    onChangeDescription = (e) => {
        const description = e.target.value;
        this.setState(() => ({
            description
        }))
    }

    onChangeNote = (e) => {
        const note = e.target.value;
        this.setState(() => ({
            note
        }))
    }

    onChangeAmount = (e) => {
        const amount = e.target.value;
        if (amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => ({
                amount
            }))
        }
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({
                createdAt
            }))
        }

    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }))
    }

    onFormSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({
                err: 'Please fill the description/amount fields.'
            }))
        }
        else {
            this.setState(() => ({
                err: ''
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 100) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf(),
            })
        }

    }
    render() {
        return (
            <div>
                {this.state.err && <p>{this.state.err}</p>}
                <form onSubmit={this.onFormSubmit}>
                    <input
                        type="text"
                        value={this.state.description}
                        onChange={this.onChangeDescription}
                        placeholder="description"
                        autoFocus
                    />
                    <input
                        type="text"
                        value={this.state.amount}
                        onChange={this.onChangeAmount}
                        placeholder="amount"
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        value={this.state.note}
                        onChange={this.onChangeNote}
                        placeholder="Add a note about your expense.">
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}