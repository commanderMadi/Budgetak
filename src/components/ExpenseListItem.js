import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

//Format currency to EGP
numeral.register('locale', 'eg', {
    delimiters: {
        thousands: ' ',
        decimal: ','
    },
    abbreviations: {
        thousand: 'k',
        million: 'm',
        billion: 'b',
        trillion: 't'
    },
    ordinal : function (number) {
        return number === 1 ? 'EGP' : 'L.E';
    },
    currency: {
        symbol: 'EGP'
    }
});



export const ExpenseListItem = ({id, description, amount, createdAt}) => (
    <div>
        <Link to={`edit/${id}`}><h3>{description}</h3></Link>
        <p>{numeral(amount / 100).format('$0,0.00')} - 
        {moment(createdAt).format('DD-MMMM-YYYY')}</p>
    </div>
)

numeral.locale('eg');

export default ExpenseListItem;