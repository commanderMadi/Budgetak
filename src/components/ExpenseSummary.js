import React from 'react';
import getVisibleExpenses from '../selectors/expenses';
import getTotalExpenses from '../selectors/total';
import { connect } from 'react-redux';
import numeral from 'numeral';

export const ExpenseSummary = (props) => {

    const totalFormatted = numeral(props.totalAmount / 100).format('$0,0.00');
    return (
        <div>

            {props.totalVisible === 1 &&<p>Viewing 1 expense totalling {totalFormatted}.</p>}

            {props.totalVisible > 1 &&<p>Viewing {props.totalVisible} expenses totalling {totalFormatted}.</p>}

        </div>
    )
}


const mapStateToProps = state => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
    return {
        totalVisible: visibleExpenses.length,
        totalAmount: getTotalExpenses(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)