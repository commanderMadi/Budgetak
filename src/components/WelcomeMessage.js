import React from 'react';
import { connect } from 'react-redux';


export class WelcomeMessage extends React.Component {
    state = {
        msgDismissed: false
    }
    constructor(props) {
        super(props);
    }
    onClickHandle = () => {
            this.setState(prevState=>({
                msgDismissed: !prevState.msgDismissed
            }));
    }

    componentDidMount() {
        const json = localStorage.getItem('dismissed');
        const msgDismissed = JSON.parse(json);
        this.setState(() => ({msgDismissed}));
    }

    componentDidUpdate() {
        if(this.state.msgDismissed) {
            const msg = JSON.stringify(this.state.msgDismissed);
            localStorage.setItem('dismissed', msg);
        }
    }
    render() {
        const startDate = this.props.filters.startDate.format('DD-MMM-YYYY');
        const endDate = this.props.filters.endDate.format('DD-MMM-YYYY');
        const userName = this.props.userName;
        return (
            <div className="content-container">
            <h1 className="welcome__username">Hello there, <span>{userName.displayName}</span>!</h1>
            {!this.state.msgDismissed &&
                <div className="welcome">
                <span id="close-msg" onClick={this.onClickHandle}>X [close message]</span>
                <p>
                By default, shown expenses are those between the start and the end
                of the current month. You can change the date range below.
                Or you can clear the date range to view a list of all your expenses
                by clicking the X button on the date range filter.
                </p>
                </div>
            }
            <p className="welcome__currently-viewing">
            You are currently viewing expenses from {startDate} to {endDate}
            </p>
            </div>
        )
    }
}


const mapStateToProps = state => ({
    filters: state.filters
})

export default connect(mapStateToProps)(WelcomeMessage);