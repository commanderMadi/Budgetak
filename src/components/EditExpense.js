import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpense extends React.Component {
    onSubmit = updates => {
        this.props.editExpense(this.props.expense.id, updates);
        this.props.history.push('/');
    }
    onClick = () => {
        this.props.startRemoveExpense(this.props.expense)
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
            This is the edit expense page.
            <ExpenseForm
                expense={this.props.expense}
                onSubmit = {this.onSubmit}
            />
            <button onClick={this.onClick}>Remove</button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id ===
    props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id,expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: expense => dispatch(startRemoveExpense(expense))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
