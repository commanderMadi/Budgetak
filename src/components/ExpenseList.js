import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectedExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="visible-narrow">Expenses</div>
            <div className="visible-large">Expense</div>
            <div className="visible-large">Amount</div>
        </div>
        <div className="list-body">
        {props.expenses.length === 0 && 
            <div className="list-item list-item--nodata">
                <span>No Expenses</span>
            </div>
            }
                {
                    props.expenses.map((expense) => {
                        return <ExpenseListItem {...expense} key={expense.id} />;
                    })
                }
        </div>
    </div>

)


const mapStateToProps = (state) => {
    return {
        expenses: selectedExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);