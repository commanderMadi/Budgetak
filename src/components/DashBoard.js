import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseSummary from './ExpenseSummary';
import ExpenseListFilters from './ExpenseListFilters';
import WelcomeMessage from './WelcomeMessage';
import { firebase } from '../firebase/firebase';

const DashBoard = () => (
    <div>
        <WelcomeMessage
        userName={firebase.auth().currentUser}
        />
        <ExpenseSummary/>
        <ExpenseListFilters/>
        <ExpenseList/>


    </div>
);

export default DashBoard;
