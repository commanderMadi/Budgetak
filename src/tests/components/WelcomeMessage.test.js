import React from 'react';
import { shallow } from 'enzyme';
import { WelcomeMessage } from '../../components/WelcomeMessage';
import { altFilters } from '../fixtures/filters';

const user = {
    displayName: 'Test user'
};

let wrapper;
beforeEach(() => {
    wrapper = shallow(<WelcomeMessage userName={user} filters={altFilters}/>);
});

test('Should render the welcome message correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should dismiss default message', () => {
    wrapper.find('#close-msg').simulate('click', ()=>{});
    expect(wrapper.state('msgDismissed')).toBe(true);
    expect(wrapper).toMatchSnapshot();
});