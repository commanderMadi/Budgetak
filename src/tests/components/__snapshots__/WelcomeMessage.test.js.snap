// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should dismiss default message 1`] = `
<div
  className="content-container"
>
  <h1
    className="welcome__username"
  >
    Hello there, 
    <span>
      Test user
    </span>
    !
  </h1>
  <p
    className="welcome__currently-viewing"
  >
    You are currently viewing expenses from 
    01-Jan-1970
     to 
    05-Jan-1970
  </p>
</div>
`;

exports[`Should render the welcome message correctly 1`] = `
<div
  className="content-container"
>
  <h1
    className="welcome__username"
  >
    Hello there, 
    <span>
      Test user
    </span>
    !
  </h1>
  <div
    className="welcome"
  >
    <span
      id="close-msg"
      onClick={[Function]}
    >
      X [close message]
    </span>
    <p>
      By default, shown expenses are those between the start and the end of the current month. You can change the date range below. Or you can clear the date range to view a list of all your expenses by clicking the X button on the date range filter.
    </p>
  </div>
  <p
    className="welcome__currently-viewing"
  >
    You are currently viewing expenses from 
    01-Jan-1970
     to 
    05-Jan-1970
  </p>
</div>
`;
